enhancements/network/user-space-pod-iface-and-api-library.md

```diff
@@ -0,0 +1,382 @@
+network devices, etc. However, these resources are presented differently inside
+the container and some of them are not visible to applications without
+explicitly injecting to the container.
+
+### Goals
+
+- Define what data needs to be passed from Kubernetes to a user space
+networking application to enable the application to run properly.
+
+- Define how this data is passed from Kubernetes to a user space networking
+application.
+
+- Ease the consumption of the data that was passed into a user space networking
+application by providing a library that both provides an example of how the
+data can be collected and if desired, integrated directly into the application. 
+
+- Ensure customers using SR-IOV can run user space networking applications in
+their deployment.
+
+### Non-Goals
+
+The following items are out of scope for this proposal:
+
+- This could, secondarily, also be used to support user space data plane
+technologies like ovs-dpdk, but its not the same, and its not our primary goal.
+
+- Non-SR-IOV devices (vDPA and vhost-user interfaces) require additional
+information. However, these devices are not required for this enhancement.
+That being said, some of the upstream work this enhancement is pulling from
+does include those interfaces. So the API Library may include some of the
+additional work for these interfaces.
+
+## Proposal
+
+As described above, there are two parts to this enhancement. Work is need to
+pass additional data to the pod to enable user space networking applications to
+run properly. Work is also needed in the pod to ease the consumption of the
+data. Each item will be described below.
+
+### Passing Data to the Pod
+
+Currently, data is passed to the pod in several forms:
+
+- **PCI Addresses**: SR-IOV Device Plugin passes the list of PCI Addresses in
+an environment variable to the pod.
+
+- **Interface Data**: Multus populates the pod annotation with a
+`k8s.v1.cni.cncf.io/network-status` field. This annotation contains the list
+of interfaces, and for each interface, the MAC address, list of IP addresses
+(not CIDR), interface (net1, net2, etc), name, etc. The pod can use the
+Kubernetes Downward API to retrieve this information.
+
+- **CPU List**: The dedicated CPU list can be retrieved by reading
+`/sys/fs/cgroup/cpuset/cpuset.cpus` in the pod.
+
+For user space networking applications running in a pod, the following
+information is still needed:
+
+- **Hugepage Information**: The type and number of hugepages allocated to the
+pod.
+
+- **Additional Interface Data**: The `k8s.v1.cni.cncf.io/network-status`
+contains a `name` field, which is the network in the
+[Network-Attachment-Definition](#networkattachmentdefinition). The `name`
+does indicate how the interface is intended to be used, however, the pod doesn't
+have enough information to determine which PCI Address is associated with which
+interface. So additional information is needed to make that association within
+the pod.
```

Comment:
I still don't understand why you're talking about `name` here... `name` is a freeform string. While it may "indicate how the interface is intended to be used" to a user, it's certainly not expected to do so in any machine-parsable way, so it seems completely irrelevant to mention it here.

Also, how do the PCI Addresses discussed here relate to the environment variables passed from the SR-IOV plugin mentioned earlier? Actually... why is the SR-IOV plugin passing information about PCI Addresses, but _not_ passing enough information to match them to interfaces? Or are we only talking about PCI Addresses in non-SR-IOV use cases here?

Reply:
So last set of questions first:
- "why is the SR-IOV plugin passing information about PCI Addresses, but not passing enough information to match them to interfaces?" Just an oversight. But there have been multiple complaints that in the container there is just a list of PCI Addresses with no additional data. Right now, if it is used at all, they are just going by the order they are in the list of PCI Addresses.
- "Or are we only talking about PCI Addresses in non-SR-IOV use cases here?" So for OCP 4.7, we are just talking SR-IOV. There is broader design to make other types of interfaces work, but OCP 4.7 is only concerned with SR-IOV.

As far as `name` just being a free-form string, "not expected to do so in any machine-parsable way" - Then that is a design flaw on my side and something else needs to be added. Currently, there is no other existing fields that I know of to convey how the interface should be used. I'll look into this more.

Reply:
Also, wrt my first bullet, SR-IOV Device Plugin does not have all the information. SR-IOV Device Plugin does not have all the config data that is in the network-attachment-definition. SR-IOV CNI has it. That is part of the NPWG new "best practices" doc. How to pass data between Plugin / CNI / Meta CNI Plugin / Container.

Reply:
So I feel like something still needs to be updated here; if the system depends on people naming their NetworkAttachmentDefinitions in a particular way, then that needs to be either fixed or explicitly documented (and then fixed later on, because, ew).

Reply:
I added a sub-bullet indicating that the "intended use" would need to be encoded in the `name` field. I will continue to work wit the upstream communities to find an alternative solution.